File: cpp-wrapper/src/SyncState.cpp
#include "../include/automerge/SyncState.hpp"

#include <utility>

namespace automerge {

// ─── Helper ───────────────────────────────────────────────────────────────────

static std::vector<uint8_t> take_cbuf_ss(uint8_t* ptr, size_t len) {
    std::vector<uint8_t> result;
    if (ptr && len > 0) {
        result.assign(ptr, ptr + len);
        AMfree_bytes(ptr, len);
    }
    return result;
}

void SyncState::check(int rc) {
    if (rc != AM_OK) {
        char buf[512]{};
        AMget_last_error(buf, sizeof(buf));
        throw AutomergeError(buf);
    }
}

// ─── Lifecycle ────────────────────────────────────────────────────────────────

SyncState::SyncState() : handle_(AMcreate_sync_state()) {
    if (!handle_) {
        throw AutomergeError("AMcreate_sync_state returned null");
    }
}

SyncState::SyncState(AMsync_state* handle) noexcept : handle_(handle) {}

SyncState::~SyncState() {
    if (handle_) {
        AMfree_sync_state(handle_);
        handle_ = nullptr;
    }
}

SyncState::SyncState(SyncState&& other) noexcept : handle_(other.handle_) {
    other.handle_ = nullptr;
}

SyncState& SyncState::operator=(SyncState&& other) noexcept {
    if (this != &other) {
        if (handle_) AMfree_sync_state(handle_);
        handle_ = other.handle_;
        other.handle_ = nullptr;
    }
    return *this;
}

// ─── Protocol ────────────────────────────────────────────────────────────────

std::vector<uint8_t> SyncState::generate_sync_message(Document& doc) {
    uint8_t* ptr{};
    size_t   len{};
    check(AMgenerate_sync_message(doc.native_handle(), handle_, &ptr, &len));
    return take_cbuf_ss(ptr, len);
}

void SyncState::receive_sync_message(Document& doc,
                                     std::span<const uint8_t> message) {
    check(AMreceive_sync_message(doc.native_handle(), handle_,
                                 message.data(), message.size()));
}

// ─── Persistence ─────────────────────────────────────────────────────────────

std::vector<uint8_t> SyncState::save() const {
    uint8_t* ptr{};
    size_t   len{};
    check(AMsave_sync_state(handle_, &ptr, &len));
    return take_cbuf_ss(ptr, len);
}

SyncState SyncState::load(std::span<const uint8_t> data) {
    AMsync_state* out{};
    int rc = AMload_sync_state(data.data(), data.size(), &out);
    if (rc != AM_OK) {
        char buf[512]{};
        AMget_last_error(buf, sizeof(buf));
        throw AutomergeError(buf);
    }
    return SyncState(out);
}

} // namespace automerge

File: csharp-wrapper/SyncState.cs
using System;

namespace Automerge.Windows
{
    /// <summary>
    /// Maintains the state of a sync session with one remote peer.
    ///
    /// Each unique peer requires its own <see cref="SyncState"/> instance.
    /// Keep it alive for the duration of the relationship so that only
    /// incremental changes are transmitted.
    /// </summary>
    public sealed class SyncState : IDisposable
    {
        private IntPtr _handle;
        private bool _disposed;

        // ─── Lifecycle ────────────────────────────────────────────────────────

        /// <summary>Create a fresh sync state.</summary>
        public SyncState()
        {
            _handle = NativeMethods.AMcreate_sync_state();
            if (_handle == IntPtr.Zero)
                throw new InvalidOperationException("Failed to create SyncState.");
        }

        private SyncState(IntPtr handle)
        {
            _handle = handle;
        }

        /// <inheritdoc />
        public void Dispose()
        {
            if (!_disposed)
            {
                if (_handle != IntPtr.Zero)
                {
                    NativeMethods.AMfree_sync_state(_handle);
                    _handle = IntPtr.Zero;
                }
                _disposed = true;
            }
        }

        // ─── Protocol ─────────────────────────────────────────────────────────

        /// <summary>
        /// Generate the next message to send to the remote peer.
        /// Returns an empty array if sync is complete.
        /// </summary>
        public byte[] GenerateSyncMessage(Document doc)
        {
            ArgumentNullException.ThrowIfNull(doc);
            ThrowIfDisposed();
            doc.ThrowIfDisposedInternal();

            IntPtr ptr = IntPtr.Zero;
            nuint len = 0;
            NativeMethods.CheckResult(
                NativeMethods.AMgenerate_sync_message(
                    doc.Handle, _handle, ref ptr, ref len));
            return NativeMethods.ReadAndFree(ptr, len);
        }

        /// <summary>
        /// Process a message received from the remote peer.
        /// </summary>
        public void ReceiveSyncMessage(Document doc, ReadOnlySpan<byte> message)
        {
            ArgumentNullException.ThrowIfNull(doc);
            ThrowIfDisposed();
            doc.ThrowIfDisposedInternal();

            unsafe
            {
                fixed (byte* m = message)
                {
                    NativeMethods.CheckResult(
                        NativeMethods.AMreceive_sync_message(
                            doc.Handle, _handle, m, (nuint)message.Length));
                }
            }
        }

        // ─── Persistence ──────────────────────────────────────────────────────

        /// <summary>Serialize the sync state for optional persistence.</summary>
        public byte[] Save()
        {
            ThrowIfDisposed();
            return NativeMethods.InvokeWithBytes(
                (ref IntPtr ptr, ref nuint len) =>
                    NativeMethods.AMsave_sync_state(_handle, ref ptr, ref len));
        }

        /// <summary>Load a sync state from persisted bytes.</summary>
        public static SyncState Load(ReadOnlySpan<byte> data)
        {
            IntPtr state = IntPtr.Zero;
            int rc;
            unsafe
            {
                fixed (byte* p = data)
                {
                    rc = NativeMethods.AMload_sync_state(p, (nuint)data.Length, ref state);
                }
            }
            NativeMethods.CheckResult(rc);
            return new SyncState(state);
        }

        /// <summary>
        /// Check if the remote peer (represented by this sync state) has all of our local changes.
        /// </summary>
        public bool HasOurChanges(Document doc)
        {
            ArgumentNullException.ThrowIfNull(doc);
            ThrowIfDisposed();
            doc.ThrowIfDisposedInternal();

            int result = 0;
            NativeMethods.CheckResult(
                NativeMethods.AMhas_our_changes(doc.Handle, _handle, ref result));
            return result != 0;
        }

        private void ThrowIfDisposed()
        {
            ObjectDisposedException.ThrowIf(_disposed, this);
        }
    }
}
